File: desktop/legacy/src/placement_managers/prop_placement_manager/handlers/prop_classifier.py
from __future__ import annotations
from typing import TYPE_CHECKING

from enums.prop_type import (
    PropType,
    big_bilateral_prop_types,
    big_unilateral_prop_types,
    small_bilateral_prop_types,
    small_unilateral_prop_types,
)
from objects.prop.prop import Prop

if TYPE_CHECKING:
    from base_widgets.pictograph.legacy_pictograph import LegacyPictograph


class PropClassifier:
    def __init__(self, pictograph: "LegacyPictograph") -> None:
        self.pictograph = pictograph
        self.big_uni: list[Prop] = []
        self.small_uni: list[Prop] = []
        self.small_bi: list[Prop] = []
        self.big_bi: list[Prop] = []
        self.hands: list[Prop] = []
        self.big_props: list[Prop] = []
        self.small_props: list[Prop] = []

    def classify_props(self) -> None:
        self.big_uni.clear()
        self.small_uni.clear()
        self.small_bi.clear()
        self.big_bi.clear()
        self.hands.clear()

        for prop in self.pictograph.elements.props.values():
            prop_type_enum = PropType.get_prop_type(prop.prop_type_str)
            if prop_type_enum in big_unilateral_prop_types:
                self.big_uni.append(prop)
            elif prop_type_enum in small_unilateral_prop_types:
                self.small_uni.append(prop)
            elif prop_type_enum in small_bilateral_prop_types:
                self.small_bi.append(prop)
            elif prop_type_enum in big_bilateral_prop_types:
                self.big_bi.append(prop)
            elif prop_type_enum == PropType.Hand:
                self.hands.append(prop)

        self.big_props = self.big_uni + self.big_bi
        self.small_props = self.small_uni + self.small_bi

File: legacy_web_app/src/lib/components/objects/Prop/PropPlacementManager/PropClassifier.ts
// PropClassifier.ts

import type { PropData } from '../PropData';
import { PropType } from './PropTypes';

const smallUnilateral = [PropType.CLUB];
const smallBilateral = [PropType.TRIAD, PropType.MINIHOOP];

export class PropClassifier {
	smallUni: any;
	smallBi: any;
	hands: any;
	bigProps: any[];
	smallProps: any[];
	bigUni: any;
	bigBi: any;

	constructor(props: PropData[]) {
		props.forEach((prop) => {
			if (smallUnilateral.includes(prop.propType)) {
				this.smallUni.push(prop);
			} else if (smallBilateral.includes(prop.propType)) {
				this.smallBi.push(prop);
			} else if (prop.propType === PropType.HAND) {
				this.hands.push(prop);
			}
		});

		this.bigProps = [...this.bigUni, ...this.bigBi];
		this.smallProps = [...this.smallUni, ...this.smallBi];
	}
}
